src/network/protocol/messages/cmpctblock.zig

```diff
@@ -0,0 +1,170 @@
+const std = @import("std");
+const protocol = @import("../lib.zig");
+
+const Sha256 = std.crypto.hash.sha2.Sha256;
+const BlockHeader = @import("../../../types/BlockHeader.zig");
+const CompactSizeUint = @import("bitcoin-primitives").types.CompatSizeUint;
+
+pub const CmpctBlockMessage = struct {
+    header: BlockHeader,
+    nonce: u64,
+    short_ids: []u64,
+    prefilled_txns: []PrefilledTransaction,
+
+    const Self = @This();
+
+    pub const PrefilledTransaction = struct {
+        index: CompactSizeUint,
+        tx: []u8, // Serialized transaction
+    };
+
+    pub fn name() *const [12]u8 {
+        return protocol.CommandNames.CMPCTBLOCK;
+    }
+
+    pub fn checksum(self: *const Self) [4]u8 {
+        var digest: [32]u8 = undefined;
+        var hasher = Sha256.init(.{});
+        self.serializeToWriter(hasher.writer()) catch unreachable;
+        hasher.final(&digest);
+
+        Sha256.hash(&digest, &digest, .{});
+
+        return digest[0..4].*;
+    }
+
+    pub fn deinit(self: *const Self, allocator: std.mem.Allocator) void {
+        allocator.free(self.short_ids);
+        for (self.prefilled_txns) |txn| {
+            allocator.free(txn.tx);
+        }
+        allocator.free(self.prefilled_txns);
+    }
+
+    pub fn serializeToWriter(self: *const Self, w: anytype) !void {
+        try self.header.serializeToWriter(w);
+        try w.writeInt(u64, self.nonce, .little);
+
+        const short_ids_count = CompactSizeUint.new(self.short_ids.len);
+        try short_ids_count.encodeToWriter(w);
+        for (self.short_ids) |id| {
+            try w.writeInt(u64, id, .little);
+        }
+
+        const prefilled_txns_count = CompactSizeUint.new(self.prefilled_txns.len);
+        try prefilled_txns_count.encodeToWriter(w);
+        for (self.prefilled_txns) |txn| {
+            try txn.index.encodeToWriter(w);
+            const tx_size = CompactSizeUint.new(txn.tx.len);
+            try tx_size.encodeToWriter(w);
+            try w.writeAll(txn.tx);
+        }
+    }
+
+    pub fn serializeToSlice(self: *const Self, buffer: []u8) !void {
+        var fbs = std.io.fixedBufferStream(buffer);
+        try self.serializeToWriter(fbs.writer());
+    }
+
+    pub fn serialize(self: *const Self, allocator: std.mem.Allocator) ![]u8 {
+        const serialized_len = self.hintSerializedLen();
+        const ret = try allocator.alloc(u8, serialized_len);
+        errdefer allocator.free(ret);
+
+        try self.serializeToSlice(ret);
+
+        return ret;
+    }
+
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !Self {
+        var cmpct_block_message: Self = undefined;
```

Comment:
only create the struct when you return. It avoid the use of undefined which is error prone